### ISSUE TITLE
code file

### ISSUE THREAD

**Issue by garvitsingh044-creator** [garvitsingh044-creator/Advanced-Satellite-Tracking-Orbit-Prediction-System#2]
import os, io, math, time, tempfile
from datetime import datetime, timedelta
import numpy as np
import pandas as pd
import requests
import streamlit as st
import plotly.graph_objects as go
from sgp4.api import Satrec, jday
from astropy.time import Time
from astropy.coordinates import get_sun, GCRS
import astropy.units as u
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, LSTM, RepeatVector, TimeDistributed, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping
import imageio
from PIL import Image
from tqdm import tqdm

st.set_page_config(layout="wide", page_title="Advanced Satellite Tracker — Reality Edition")

CELESTRAK_BASE = "https://celestrak.org/NORAD/elements/"
CELESTRAK_LISTS = {
    "stations": CELESTRAK_BASE + "stations.txt",
    "visual": CELESTRAK_BASE + "visual.txt",
    "active": CELESTRAK_BASE + "active.txt",
    "starlink": CELESTRAK_BASE + "starlink.txt",
}

def fetch_tle_list(url, timeout=15):
    r = requests.get(url, timeout=timeout)
    r.raise_for_status()
    lines = [ln.strip() for ln in r.text.splitlines() if ln.strip() != ""]
    blocks = []
    for i in range(0, len(lines)-2, 3):
        blocks.append((lines[i], lines[i+1], lines[i+2]))
    return blocks

def propagate_tle(line1, line2, times_utc):
    sat = Satrec.twoline2rv(line1, line2)
    rows=[]
    for t in times_utc:
        jd, fr = jday(t.year, t.month, t.day, t.hour, t.minute, t.second + t.microsecond*1e-6)
        e, r, v = sat.sgp4(jd, fr)
        if e != 0:
            rows.append([t, np.nan,np.nan,np.nan,np.nan,np.nan,np.nan])
        else:
            rows.append([t, r[0]*1000., r[1]*1000., r[2]*1000., v[0]*1000., v[1]*1000., v[2]*1000.])
    df = pd.DataFrame(rows, columns=['time_utc','x_m','y_m','z_m','vx_mps','vy_mps','vz_mps'])
    return df

def build_uniform_times(start_dt, duration_hours, dt_seconds):
    steps = int((duration_hours*3600)/dt_seconds) + 1
    return [start_dt + timedelta(seconds=i*dt_seconds) for i in range(steps)]

def sun_vector_eci(dt_utc):
    t = Time(dt_utc, scale='utc')
    sun_icrs = get_sun(t)
    sun_gcrs = sun_icrs.transform_to(GCRS(obstime=t))
    cart = sun_gcrs.cartesian
    vec = np.array([cart.x.to(u.m).value, cart.y.to(u.m).value, cart.z.to(u.m).value])
    return vec

def atmospheric_density(h_km, f107=150):
    rho0 = 1.225
    rho = 1e-9 * np.exp(-h_km/50.0) * (1 + (f107-150)/500.0)
    return rho

class SimpleCVKalman:
    def __init__(self, dt, process_var=1.0, meas_var=100.0):
        self.dt = dt
        self.F = np.eye(6)
        for i in range(3):
            self.F[i, i+3] = dt
        q = process_var
        self.Q = q * np.eye(6)
        r = meas_var
        self.R = r * np.eye(3)
        self.P = np.eye(6) * 1e3
        self.x = np.zeros(6)
        self.initialized = False
    def init(self, pos, vel=None):
        self.x[:3] = pos
        self.x[3:] = vel if (vel is not None) else 0.0
        self.initialized = True
    def predict(self):
        self.x = self.F @ self.x
        self.P = self.F @ self.P @ self.F.T + self.Q
    def update(self, z):
        H = np.zeros((3,6)); H[0,0]=1; H[1,1]=1; H[2,2]=1
        S = H @ self.P @ H.T + self.R
        K = self.P @ H.T @ np.linalg.inv(S)
        y = z - H @ self.x
        self.x = self.x + K @ y
        self.P = (np.eye(6) - K @ H) @ self.P
    def step(self, meas):
        if not self.initialized:
            self.init(meas, vel=np.zeros(3))
        self.predict()
        self.update(meas)
        return self.x.copy()

def build_seq2seq_resid(seq_len, n_features, pred_steps, units=128, dropout_rate=0.2):
    inp = Input(shape=(seq_len, n_features))
    x = LSTM(units, return_sequences=False)(inp)
    x = Dropout(dropout_rate)(x, training=True)
    x = RepeatVector(pred_steps)(x)
    x = LSTM(units, return_sequences=True)(x)
    x = TimeDistributed(Dense(64, activation='relu'))(x)
    out = TimeDistributed(Dense(3))(x)
    model = Model(inp, out)
    model.compile(optimizer='adam', loss='mse')
    return model

def mc_predict(model, X, mc_runs=30):
    preds = []
    for _ in range(mc_runs):
        preds.append(model.predict(X, verbose=0))
    arr = np.stack(preds, axis=0)
    mean = arr.mean(axis=0)
    std = arr.std(axis=0)
    return mean, std

st.title("Advanced Satellite Tracker ")
st.markdown("Hybrid physics + ML, multi-satellite, day/night shading, starfield, uncertainty, KF smoothing, MP4 export.")

col1, col2 = st.columns([1,2])

with col1:
    st.subheader("Data & propagation")
    list_choice = st.selectbox("Choose CelesTrak list", options=list(CELESTRAK_LISTS.keys()), index=0)
    try:
        tle_blocks = fetch_tle_list(CELESTRAK_LISTS[list_choice])
    except Exception as e:
        st.error(f"Could not fetch TLE list: {e}")
        tle_blocks = []
    sat_options = [b[0] for b in tle_blocks[:200]]
    sat_selected = st.multiselect("Choose satellites (max 6)", options=sat_options, default=sat_options[:1], max_selections=6)
    duration_hours = st.number_input("Propagation duration (hours)", min_value=0.5, max_value=24.0, value=3.0, step=0.5)
    dt_seconds = st.number_input("Sample dt (s)", min_value=10, max_value=300, value=30, step=10)
    start_now = st.checkbox("Propagate from now (UTC)", value=True)
    if start_now:
        start_dt = datetime.utcnow()
    else:
        start_dt = st.date_input("Start date (UTC)", value=datetime.utcnow().date())
        start_dt = datetime.combine(start_dt, datetime.min.time())

    build_button = st.button("Propagate selected satellites")

with col2:
    st.subheader("Model & Visualization")
    seq_len = st.slider("Sequence length (timesteps)", min_value=5, max_value=60, value=30, step=5)
    pred_steps = st.slider("Prediction horizon (timesteps)", min_value=1, max_value=30, value=10, step=1)
    train_epochs = st.slider("Train epochs (small for demo)", min_value=1, max_value=100, value=12, step=1)
    use_kf = st.checkbox("Apply Kalman filter smoothing to predictions", value=True)
    use_atm_drag = st.checkbox("Attempt simple atmospheric drag correction (placeholder)", value=False)
    mc_runs = st.slider("MC dropout runs for uncertainty", min_value=5, max_value=100, value=30, step=5)
    exaggeration = st.slider("Radial exaggeration for visualization", min_value=1.0, max_value=6.0, value=2.0, step=0.1)
    trail_width = st.slider("Trail width", min_value=1, max_value=8, value=3, step=1)
    export_mp4 = st.button("Export MP4 (renders frames)")

if build_button and len(sat_selected) > 0:
    st.info("Propagating and preparing dataset... this may take a moment.")
    selected_blocks = []
    for name in sat_selected:
        for b in tle_blocks:
            if b[0] == name:
                selected_blocks.append(b)
                break

    times = build_uniform_times(start_dt, duration_hours, dt_seconds)

    propagated = {}
    for name, l1, l2 in selected_blocks:
        df_pos = propagate_tle(l1, l2, times)
        propagated[name] = df_pos
        st.write(f"{name}: propagated {len(df_pos)} rows")

    results = {}
    progress_bar = st.progress(0)
    total = len(selected_blocks)
    for idx, (name, l1, l2) in enumerate(selecteded_blocks if False else selected_blocks):
        df = propagated[name].dropna().reset_index(drop=True)
        df['r_m'] = np.sqrt(df.x_m**2 + df.y_m**2 + df.z_m**2)
        df['speed_mps'] = np.sqrt(df.vx_mps**2 + df.vy_mps**2 + df.vz_mps**2)
        features = ['x_m','y_m','z_m','vx_mps','vy_mps','vz_mps','r_m','speed_mps']
        X, Y = [], []
        N = len(df)
        for i in range(seq_len, N - pred_steps):
            X.append(df.loc[i-seq_len:i-1, features].values)
            Y.append(df.loc[i:i+pred_steps-1, ['x_m','y_m','z_m']].values)
        X = np.array(X); Y = np.array(Y)
        if len(X) < 20:
            st.warning(f"Not enough samples for {name} (need >20); skipping training but we will still plot propagated path.")
            results[name] = {'df': df, 'trained': False}
            progress_bar.progress((idx+1)/total)
            continue

        train_end = int(0.8 * len(X))
        X_train, X_test = X[:train_end], X[train_end:]
        Y_train, Y_test = Y[:train_end], Y[train_end:]

        scaler_X = StandardScaler()
        scaler_X.fit(X_train.reshape(-1, X_train.shape[2]))
        X_train_s = scaler_X.transform(X_train.reshape(-1, X_train.shape[2])).reshape(X_train.shape)
        X_test_s  = scaler_X.transform(X_test.reshape(-1, X_test.shape[2])).reshape(X_test.shape)

        def baseline_multi_from_df_at_index(df_local, idx_global):
            out = []
            for k in range(1, pred_steps+1):
                tt = df_local.loc[idx_global + k, ['x_m','y_m','z_m']].values
                out.append(tt)
            return np.array(out)
        baseline_test = []
        for i_sample in range(train_end, len(X)):
            orig_idx = seq_len + i_sample
            baseline_test.append(df.loc[orig_idx:orig_idx+pred_steps-1, ['x_m','y_m','z_m']].values)
        baseline_test = np.array(baseline_test)

        resid_train = Y_train - np.stack([df.loc[i: i+pred_steps-1, ['x_m','y_m','z_m']].values for i in range(seq_len, seq_len + len(X_train))])
        resid_scaler = StandardScaler()
        resid_scaler.fit(resid_train.reshape(-1,3))
        resid_train_s = resid_scaler.transform(resid_train.reshape(-1,3)).reshape(resid_train.shape)

        model = build_seq2seq_resid(seq_len, X.shape[2], pred_steps, units=128, dropout_rate=0.2)
        es = EarlyStopping(monitor='val_loss', patience=5, restore_best_weights=True)
        baseline_train = []
        for i_sample in range(seq_len, seq_len + len(X_train)):
            baseline_train.append(df.loc[i_sample: i_sample+pred_steps-1, ['x_m','y_m','z_m']].values)
        baseline_train = np.array(baseline_train)
        resid_train = Y_train - baseline_train
        resid_train_s = resid_scaler.transform(resid_train.reshape(-1,3)).reshape(resid_train.shape)

        history = model.fit(X_train_s, resid_train_s, validation_split=0.1, epochs=train_epochs, batch_size=64, callbacks=[es], verbose=0)
        st.write(f"Trained residual model for {name} — final val_loss: {history.history['val_loss'][-1]:.6f}")

        mean_resid_s, std_resid_s = mc_predict(model, X_test_s, mc_runs=mc_runs)
        mean_resid = resid_scaler.inverse_transform(mean_resid_s.reshape(-1,3)).reshape(mean_resid_s.shape)
        std_resid = resid_scaler.scale_ * std_resid_s
        pred_hybrid = baseline_test + mean_resid

        if use_kf:
            kf = SimpleCVKalman(dt=dt_seconds, process_var=1.0, meas_var=500.0)
            smoothed = []
            for i_sample in range(pred_hybrid.shape[0]):
                seq_sm = []
                last_obs_idx = seq_len + (train_end + i_sample) - 1
                last_pos = df.loc[last_obs_idx, ['x_m','y_m','z_m']].values
                last_vel = df.loc[last_obs_idx, ['vx_mps','vy_mps','vz_mps']].values
                kf.init(last_pos, last_vel)
                for h in range(pred_steps):
                    pred = pred_hybrid[i_sample, h]
                    kf_pred = kf.step(pred[:3])
                    seq_sm.append(kf_pred[:3])
                smoothed.append(np.array(seq_sm))
            pred_hybrid_smoothed = np.array(smoothed)
        else:
            pred_hybrid_smoothed = pred_hybrid

        true_test = Y_test
        rmse_first = np.sqrt(np.mean(np.sum((true_test[:,0,:] - pred_hybrid_smoothed[:,0,:])**2, axis=1)))
        st.write(f"{name} first-step hybrid RMSE (meters): {rmse_first:.1f}")

        results[name] = {
            'df': df,
            'model': model,
            'scaler_X': scaler_X,
            'resid_scaler': resid_scaler,
            'pred_hybrid': pred_hybrid,
            'pred_hybrid_smoothed': pred_hybrid_smoothed,
            'baseline_test': baseline_test,
            'X_test': X_test,
            'Y_test': Y_test,
            'trained': True,
            'rmse_first': rmse_first
        }
        progress_bar.progress((idx+1)/total)
    st.success("Propagation & training complete.")

    st.subheader("Interactive 3D Viewer")
    vis_sat = st.selectbox("Choose satellite to view", options=list(results.keys()))
    chosen = results[vis_sat]
    df = chosen['df']
    path = df[['x_m','y_m','z_m']].values
    if chosen['trained']:
        preds = chosen['pred_hybrid_smoothed']
        test_start_global_idx = seq_len + int(0.8 * (len(df) - seq_len - pred_steps))
        times_plot = df['time_utc'].astype(str).tolist()
    else:
        preds = None

    nstars = 1500
    rng = np.random.default_rng(1234)
    stars = rng.uniform(-2e7, 2e7, size=(nstars,3))

    R_earth = 6378137.0
    u = np.linspace(0, 2*np.pi, 60)
    v = np.linspace(0, np.pi, 30)
    ex = (R_earth * np.outer(np.cos(u), np.sin(v))).flatten()
    ey = (R_earth * np.outer(np.sin(u), np.sin(v))).flatten()
    ez = (R_earth * np.outer(np.ones_like(u), np.cos(v))).flatten()

    fig = go.Figure()
    fig.add_trace(go.Scatter3d(x=stars[:,0], y=stars[:,1], z=stars[:,2],
                               mode='markers', marker=dict(size=1, color='white', opacity=0.8),
                               name='stars'))
    fig.add_trace(go.Scatter3d(x=ex, y=ey, z=ez, mode='markers', marker=dict(size=1,color='royalblue', opacity=0.7), name='Earth'))

    fig.add_trace(go.Scatter3d(x=path[:,0], y=path[:,1], z=path[:,2],
                               mode='lines', line=dict(width=2, color='cyan'), name=f'{vis_sat} propagated'))

    if preds is not None:
        nshow = min(200, preds.shape[0])
        train_end = int(0.8 * (len(df) - seq_len - pred_steps))
        base_idx = seq_len + int(0.8 * (len(df)-seq_len-pred_steps))
        pred_points = preds[:nshow,0,:]
        fig.add_trace(go.Scatter3d(x=pred_points[:,0], y=pred_points[:,1], z=pred_points[:,2],
                                   mode='markers', marker=dict(size=3+trail_width, color='gold'), name='Hybrid pred (h=1)'))

    try:
        sun_vec = sun_vector_eci(datetime.utcnow())
        dots = (path @ sun_vec)
        bright_mask = dots > 0
        fig.add_trace(go.Scatter3d(x=path[bright_mask,0], y=path[bright_mask,1], z=path[bright_mask,2],
                                   mode='markers', marker=dict(size=1.5, color='yellow', opacity=0.6), name='Day-side'))
    except Exception:
        pass

    fig.update_layout(scene=dict(xaxis=dict(showbackground=False, visible=False),
                                 yaxis=dict(showbackground=False, visible=False),
                                 zaxis=dict(showbackground=False, visible=False)),
                      paper_bgcolor='black', plot_bgcolor='black', legend=dict(bgcolor='rgba(0,0,0,0.4)', font=dict(color='white')))

    st.plotly_chart(fig, use_container_width=True, height=700)

    if export_mp4:
        st.info("Exporting MP4 — rendering frames (this can take time).")
        frames = 240
        out_path = os.path.join("outputs_reality", f"{vis_sat}_render_{int(time.time())}.mp4")
        os.makedirs("outputs_reality", exist_ok=True)
        tmp_frames = []
        for fr in tqdm(range(frames), desc="Rendering frames"):
            az = fr * (360/frames)
            elev = 20 + 8 * math.sin(fr / 80.0)
            camera = dict(eye=dict(x=2.2*R_earth*math.cos(math.radians(az)), y=2.2*R_earth*math.sin(math.radians(az)), z=R_earth*0.3*math.sin(math.radians(elev))))
            fig.update_layout(scene_camera=camera)
            img_bytes = fig.to_image(format="png", width=1280, height=720, scale=1)
            tmp_frames.append(imageio.v2.imread(img_bytes))
        imageio.mimsave(out_path, tmp_frames, fps=24, codec='libx264')
        st.success(f"Saved MP4 to {out_path}. You can download it from the server.")

    if chosen.get('trained', False):
        sample_N = min(200, chosen['pred_hybrid_smoothed'].shape[0])
        out_rows = []
        for i in range(sample_N):
            out_rows.append({
                'true_x': float(chosen['Y_test'][i,0,0]),
                'true_y': float(chosen['Y_test'][i,0,1]),
                'true_z': float(chosen['Y_test'][i,0,2]),
                'pred_x': float(chosen['pred_hybrid_smoothed'][i,0,0]),
                'pred_y': float(chosen['pred_hybrid_smoothed'][i,0,1]),
                'pred_z': float(chosen['pred_hybrid_smoothed'][i,0,2]),
            })
        outdf = pd.DataFrame(out_rows)
        fname = f"{vis_sat.replace(' ','_')}_sample_compare.csv"
        outdf.to_csv(fname, index=False)
        st.download_button("Download sample CSV (first-step)", data=outdf.to_csv(index=False).encode('utf-8'), file_name=fname)

    st.write("Done. You can train other satellites or export another MP4.")

else:
    st.info("Choose satellites and press 'Propagate selected satellites' to run propagation, training, and visualization. For a demo, choose ISS (Stations list).")
